lib/Optimizer/Transforms/DependencyAnalysis.cpp

```diff
@@ -0,0 +1,2561 @@
+/*******************************************************************************
+ * Copyright (c) 2022 - 2024 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#include "PassDetails.h"
+#include "cudaq/Frontend/nvqpp/AttributeNames.h"
+#include "cudaq/Optimizer/Dialect/CC/CCOps.h"
+#include "cudaq/Optimizer/Dialect/Quake/QuakeDialect.h"
+#include "cudaq/Optimizer/Dialect/Quake/QuakeTypes.h"
+#include "cudaq/Optimizer/Transforms/Passes.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/Passes.h"
+
+#define DEBUG_TYPE "dep-analysis"
+
+using namespace mlir;
+
+#define RAW(X) quake::X
+#define RAW_MEASURE_OPS MEASURE_OPS(RAW)
+#define RAW_GATE_OPS GATE_OPS(RAW)
+#define RAW_QUANTUM_OPS QUANTUM_OPS(RAW)
+
+//===----------------------------------------------------------------------===//
+// Generated logic
+//===----------------------------------------------------------------------===//
+namespace cudaq::opt {
+#define GEN_PASS_DEF_DEPENDENCYANALYSIS
+#include "cudaq/Optimizer/Transforms/Passes.h.inc"
+} // namespace cudaq::opt
+
+namespace {
+typedef size_t PhysicalQID;
+typedef size_t VirtualQID;
+
+size_t getOperandIDXFromResultIDX(size_t resultidx, Operation *op) {
+  if (isa<RAW_MEASURE_OPS>(op))
+    return 0;
+  if (isa<quake::SwapOp>(op))
+    return (resultidx == 0 ? 1 : 0);
```

Comment:
Why is the swap gate not treated like any other 2-qubit gate? Is this attempting to reconstruct something from swap gates that you have inserted? If not, then I think this should simply fall through into the for loop below.

Reply:
I think this was a mis-guided attempt to get my pass to play nicely with measurement ordering, so I don't think it's a necessarily thing to keep (and actually could lead to incorrect behavior even).

Reply:
Fixed